src/eachkmer.jl

```diff
@@ -35,71 +67,140 @@ end
-            isok &= iscertain(nt)
-        else
-            for i in 1:it.step
-                nt = inbounds_getindex(it.seq, pos+i-1+offset)
-                kmer = kmer << 2 | trailing_zeros(nt)
-                isok &= iscertain(nt)
-            end
-        end
-        if isok
-            return (pos, T(kmer)), (pos+it.step, kmer)
+    while i ≤ it.stop
+        nt = reinterpret(Int8, inbounds_getindex(it.seq, i))
+        @inbounds val = kmerbits[nt + 1]
+        kmer = kmer << 2 | val
+        filled += 1
+        if filled == kmersize(T)
+            return (1, T(kmer)), (i, kmer)
         end
+        i += 1
     end
+    return nothing
+end
+
+function Base.iterate(it::EveryKmerIterator{T,S}, state
+    ) where {T,S<:BioSequence{<:TwoBitNucs}}
+    i, kmer = state
+    i += 1
+
+    if i > it.stop
+        return nothing
+    else
+        nt = reinterpret(Int8, inbounds_getindex(it.seq, i))
+        @inbounds val = kmerbits[nt + 1]
+        kmer = kmer << 2 | val
+        pos = i - kmersize(T) + 1
+        return (pos, T(kmer)), (i, kmer)
+    end
+end
 
-    while pos + k - 1 ≤ lastindex(it.seq)
-        kmer, ok = extract_kmer_impl(it.seq, pos, k)
-        if ok
-            return (pos, T(kmer)), (pos+it.step, kmer)
+function Base.iterate(it::SpacedKmerIterator{T,S}, state
+    ) where {T,S<:BioSequence{<:TwoBitNucs}}
+    i, kmer = state
+    filled = it.filled
+    i += it.increment
+
+    while i ≤ it.stop
+        nt = reinterpret(Int8, inbounds_getindex(it.seq, i))
+        @inbounds val = kmerbits[nt + 1]
+        kmer = kmer << 2 | val
+        filled += 1
+        if filled == kmersize(T)
+            pos = i - kmersize(T) + 1
+            return (pos, T(kmer)), (i, kmer)
         end
-        pos += it.step
+        i += 1
     end
+    return nothing
+end
+
+function Base.iterate(it::EveryKmerIterator{T,S}, state=(it.start-1,1,UInt64(0))
+    ) where {T,S<:Union{ReferenceSequence, BioSequence{<:FourBitNucs}}}
+    i, filled, kmer = state
+    i += 1
+    filled -= 1
 
+    while i ≤ it.stop
+        nt = reinterpret(Int8, inbounds_getindex(it.seq, i))
+        @inbounds val = kmerbits[nt + 1]
+        kmer = kmer << 2 | val
+        filled = ifelse(val == 0xff, 0, filled+1)
+
+        if filled == kmersize(T)
+            pos = i - kmersize(T) + 1
+            return (pos, T(kmer)), (i, kmersize(T), kmer)
+        end
+        i += 1
+    end
     return nothing
 end
 
-function extract_kmer_impl(seq, from, k)
-    kmer::UInt64 = 0
-    isok = true
-    for i in 1:k
-        nt = inbounds_getindex(seq, from+i-1)
-        kmer = kmer << 2 | trailing_zeros(nt)
-        isok &= iscertain(nt)
+@inline function Base.iterate(it::SpacedKmerIterator{T,S}, state=(it.start-it.increment, 1, 0, UInt64(0))
+    ) where {T,S<:Union{ReferenceSequence, BioSequence{<:FourBitNucs}}}
+    i, pos, filled, kmer = state
+    i += it.increment
+
+    while i ≤ it.stop
+        nt = reinterpret(Int8, inbounds_getindex(it.seq, i))
```

Comment:
Should this be UInt8?

Reply:
Sure, it compiles to the same instructions anyway, so shouldn't matter.